File: gospider/downloader/downloader.go
package downloader

import (
	"io"
	"net/http"
	"net/http/cookiejar"
	"time"
)

// Downloader 下载器接口
type Downloader interface {
	Download(req *Request) *Response
}

// Request 请求对象（避免循环依赖）
type Request struct {
	URL     string
	Method  string
	Headers map[string]string
	Body    io.Reader
}

// Response 响应对象（避免循环依赖）
type Response struct {
	URL        string
	StatusCode int
	Headers    http.Header
	Body       []byte
	Text       string
	Request    *Request
	Duration   time.Duration
	Error      error
}

// HTTPDownloader HTTP 下载器
type HTTPDownloader struct {
	client *http.Client
}

// NewDownloader 创建下载器
func NewDownloader() *HTTPDownloader {
	jar, _ := cookiejar.New(nil)
	return &HTTPDownloader{
		client: &http.Client{
			Timeout: 30 * time.Second,
			Jar:     jar,
		},
	}
}

// Download 下载页面
func (d *HTTPDownloader) Download(req *Request) *Response {
	startTime := time.Now()

	// 创建 HTTP 请求
	httpReq, err := http.NewRequest(req.Method, req.URL, req.Body)
	if err != nil {
		return &Response{
			URL:   req.URL,
			Error: err,
		}
	}

	// 设置请求头
	for key, value := range req.Headers {
		httpReq.Header.Set(key, value)
	}

	// 执行请求
	httpResp, err := d.client.Do(httpReq)
	duration := time.Since(startTime)

	if err != nil {
		return &Response{
			URL:      req.URL,
			Error:    err,
			Duration: duration,
		}
	}
	defer httpResp.Body.Close()

	// 读取响应体
	body, err := io.ReadAll(httpResp.Body)
	if err != nil {
		return &Response{
			URL:      req.URL,
			Error:    err,
			Duration: duration,
		}
	}

	return &Response{
		URL:        req.URL,
		StatusCode: httpResp.StatusCode,
		Headers:    httpResp.Header,
		Body:       body,
		Text:       string(body),
		Request:    req,
		Duration:   duration,
		Error:      nil,
	}
}

// SetTimeout 设置超时
func (d *HTTPDownloader) SetTimeout(timeout time.Duration) {
	d.client.Timeout = timeout
}

// SetCookie 设置 Cookie
func (d *HTTPDownloader) SetCookie(domain string, cookies map[string]string) {
	// Cookie 处理逻辑
}

File: rustspider/src/downloader.rs
//! HTTP 下载器

use crate::models::{Request, Response};
use std::collections::HashMap;
use std::sync::Arc;
use std::time::{Duration, Instant};

/// HTTP 下载器
#[derive(Clone)]
pub struct HTTPDownloader {
    client: Arc<reqwest::blocking::Client>,
}

impl HTTPDownloader {
    /// 创建新下载器
    pub fn new() -> Self {
        let client = reqwest::blocking::Client::builder()
            .timeout(Duration::from_secs(30))
            .build()
            .unwrap();

        HTTPDownloader {
            client: Arc::new(client),
        }
    }

    /// 下载页面
    pub fn download(&self, req: &Request) -> Response {
        let start = Instant::now();

        // 创建请求
        let mut req_builder = match req.method.as_str() {
            "POST" => self.client.post(&req.url),
            "PUT" => self.client.put(&req.url),
            "DELETE" => self.client.delete(&req.url),
            _ => self.client.get(&req.url),
        };

        // 设置请求头
        for (key, value) in &req.headers {
            req_builder = req_builder.header(key, value);
        }

        // 设置请求体
        if let Some(body) = &req.body {
            req_builder = req_builder.body(body.clone());
        }

        // 执行请求
        match req_builder.send() {
            Ok(resp) => {
                let status_code = resp.status().as_u16();
                let headers = resp
                    .headers()
                    .iter()
                    .map(|(k, v)| (k.to_string(), v.to_str().unwrap_or("").to_string()))
                    .collect();

                let bytes = resp.bytes().unwrap_or_default();
                let text = String::from_utf8_lossy(&bytes).to_string();

                Response {
                    url: req.url.clone(),
                    status_code,
                    headers,
                    text,
                    bytes: bytes.to_vec(),
                    duration: start.elapsed(),
                    error: None,
                }
            }
            Err(e) => Response {
                url: req.url.clone(),
                status_code: 0,
                headers: HashMap::new(),
                text: String::new(),
                bytes: Vec::new(),
                duration: start.elapsed(),
                error: Some(e.to_string()),
            },
        }
    }

    /// 设置超时
    pub fn with_timeout(timeout_secs: u64) -> Self {
        let client = reqwest::blocking::Client::builder()
            .timeout(Duration::from_secs(timeout_secs))
            .build()
            .unwrap();

        HTTPDownloader {
            client: Arc::new(client),
        }
    }
}

impl Default for HTTPDownloader {
    fn default() -> Self {
        Self::new()
    }
}
